Initial commit of support for Google Compute Engine.

wrigri opened the pull request:
The unit tests are still in development and will be coming in a separate commit.

This currently works, and does not rely on any additional third-party packages, with one exception:  If you want to use the "Service Account" OAUTH authentication, you will need the PyCrypto package installed.  However, it fails gracefully if it's not there.

I'm particularly interested to see if you notice any places where I broke the established API.  I believe that I followed it as closely as possible, however.

wrigri committed:
Initial commit of support for Google Compute Engine.
The unit tests are still in development and will be coming in a separate commit.

wrigri commented:
Also, I should note that this has currently only been fully tested with Python 2.7.  It's likely that some minor adjustments will need to be made for Python 3.X (and maybe 2.5 if 2.5 support isn't officially dropped before this is merged in).

Kami commented on demos/secrets.py-dist:
```diff
@@ -22,6 +22,10 @@ DREAMHOST_PARAMS = ('key',)
 GANDI_PARAMS = ('user',)
+GCE_PARAMS = ('email_address', 'key') # Service Account Authentication
+#GCE_PARAMS = ('client_id', 'client_secret') # Installed App Authentication
+GCE_KEYWORD_PARAMS = ({'project': 'project_name', 'ssh_username': None,
```
Parenthesis are unnecessary around the dict which is defined over multiple lines.

wrigri replied:
Fixed.

Kami commented on libcloud/common/google.py:
```diff
@@ -0,0 +1,466 @@
+        @return:  A dictionary with updated token information
+        @rtype:   C{dict}
+        """
+        data = urllib.urlencode(request_body)
```
You need to use `urlencode` from `libcloud.utils.py3` (https://github.com/apache/libcloud/blob/trunk/libcloud/utils/py3.py), otherwise it won't work will all the supported versions

wrigri replied:
Thanks for pointing that out.  I made the changes for urlencode and also for urlparse.

Kami commented on libcloud/common/google.py:
```diff
@@ -0,0 +1,466 @@
+        if self.login_hint:
+            auth_params['login_hint'] = self.login_hint
+
+        data = urllib.urlencode(auth_params)
```
Same here (use `urlencode` from `libcloud.utils.py3`.

wrigri replied:
Fixed

Kami commented on libcloud/common/google.py:
```diff
@@ -0,0 +1,466 @@
+        if SHA256 is None:
+            raise GoogleAuthError('PyCrypto library required for '
+                                  'Service Accout Authentication.')
+        keypath = os.path.expanduser(key)
```
I'm not a big fan of a single argument containing two different type of values (either a path or a string), but if we proceed with this approach I would rather check if `key` is a file file path like this:

``` python
is_file_path = os.path.exists(keypath) and os.path.isfile(keypath)
...
```

wrigri replied:
OK, I updated the check.  I don't have a strong opinion about the best way to pass this key in, so I'm certainly open to suggestions about the best way to handle it.  The key file itself is a little less than 1KB in size, so copying and pasting it as a string it a little cumbersome.  I would assume that most people would want to read it out of a file.  Of course, we could leave it to the end user to read the file themselves and pass the string.  Anyway, like I said, I'm open to suggestions about the UI for this.

Kami commented on libcloud/common/google.py:
```diff
@@ -0,0 +1,466 @@
+                                  'Service Accout Authentication.')
+        keypath = os.path.expanduser(key)
+        try:
+            key = open(keypath).read()
```
Please use `with` context manager when opening and reading a file (e.g. https://github.com/apache/libcloud/blob/trunk/libcloud/compute/deployment.py#L188).

This way you can be sure file is always closed (even if the exception happens) and you don't need to manually wrap it in `try / except / finally` block and make sure you always close file in the `finally` block.

wrigri replied:
Done. (I prefer 'with' myself, but wasn't sure if you were using it with your 2.5 support.)

Kami replied:
Thanks.

`with` is also available in Python 2.5, you just need to add `from __future__ import with_statement` to the top of the file where you use `with` (e.g. https://github.com/apache/libcloud/blob/trunk/libcloud/test/file_fixtures.py#L17)

Kami commented on libcloud/common/google.py:
```diff
@@ -0,0 +1,466 @@
+        token_info = None
+        filename = os.path.realpath(os.path.expanduser(self.credential_file))
+        try:
+            f = open(filename)
```
Same here, please use `with`.

wrigri replied:
Done.

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -0,0 +1,1854 @@
+        """
+        Returns the next Maintenance Window.
+
+        @return:  A dictionary containing maintenance window info
```
Would be nice to document which keys returned value contains because now just by looking at the docstring I have no idea what kind of dictionary this function returns.

wrigri replied:
Done.

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -0,0 +1,1854 @@
+        Returns the duration of the next maintenance window.
+
+        @return:  Duration of next maintenance window
+        @rtype:   C{datetime.timedelta}
```
Cool, I like that you are using `timedelta` instead of seconds.

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -0,0 +1,1854 @@
+        if project:
+            self.project = project
+        else:
+            print "Please specify the project in your Driver's constructor"
```
Please throw `ValueError` instead of exiting here.

wrigri replied:
Done

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -0,0 +1,1854 @@
+        """
+        request = '/aggregated/%s' % res_type
+        res_list = self.connection.request(request).object
+        for k, v in res_list['items'].items():
```
We could probably use multiple list comprehensions to make this slightly more readable and remove multiple levels of nested.

wrigri replied:
OK, I cleaned things up a bit, replaced several for loops with list comprehensions and for the more complex ones like this one, got it down to a single for loop.

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -0,0 +1,1854 @@
+        if 'items' in response:
+            # The aggregated result returns dictionaries for each region
+            if region is None:
+                for k, v in response['items'].items():
```
Same here (reducing nesting)

wrigri replied:
Done.

Kami commented:
I've glanced over the code and posted some comments in-line. There are some issues, but it shouldn't be too hard to address those:
- Making sure it works with all the supported Python versions (after you add the tests it should be easier to test that using `tox`)
- Some code readability can be improved by making the code more flat and reducing the levels of nesting (either by using multiple list comprehensions or by splitting code in multiple functions)

As far as the high level stuff goes - overall all the class relationships and driver look solid.

wrigri committed:
Address specific issues highlighted in github review.
Includes:
Using the with statement
Using urlencode and urlparse from libcloud.utils.py3
Clean-up of nested for loops
Some additional documentation
Other minor fixes

Kami commented on libcloud/common/google.py:
```diff
@@ -0,0 +1,467 @@
+                return super(GoogleBaseConnection, self).request(
+                    *args, **kwargs)
+            except socket.error as e:
+                if e.errno == 104:
```
Please use a constant - `errno.ECONNRESET`

Kami commented on libcloud/common/google.py:
```diff
@@ -0,0 +1,467 @@
+            try:
+                return super(GoogleBaseConnection, self).request(
+                    *args, **kwargs)
+            except socket.error as e:
```
Need to use this way of extracting an exception - https://github.com/apache/libcloud/blob/trunk/libcloud/utils/misc.py#L68

Otherwise it won't work with all the supported Python versions.

Kami commented on libcloud/common/google.py:
```diff
@@ -0,0 +1,467 @@
+                if e.errno == 104:
+                    tries = tries + 1
+                else:
+                    raise
```
Please raise the original exception - `raise e`

Kami commented:
@wrigri Spotted a couple of minor things, otherwise looks good.

Should be ready to be merged once you finish the tests.

wrigri committed:
Fixes for handling socket error.

wrigri committed:
Rewrite reboot_node.
Google just added reboot support to the API, so use the API instead of SSH.

wrigri committed:
Various minor fixes that were discovered during unittest development.

wrigri committed:
Update demo script with a couple extra features and make it importable.
Also, update secrets.py-dist to remove the ssh keywords.

wrigri committed:
Fixtures for Google Compute Engine Tests

wrigri committed:
Misc PEP8 fixes.

wrigri committed:
Additional Documentation, PEP8 fixes, and PY3 fixes.

wrigri committed:
Add unittests for Google Compute Engine

wrigri commented:
OK, I've added unittests.  I've tested (unittests and integration test (gce_demo.py)) against Python 2.7 and Python 3.2.  I'll run tox on the whole thing tomorrow and make sure there aren't any other cross-version issues that need to be addressed.

Other than that, the code should all be PEP8 compliant (except for some long lines due to long method names in test_gce.py), and epydoc doesn't show any errors (except that it complains about 'inherits').

wrigri committed:
Remove keywords for SSH information, since it's no longer needed for reboot.

Kami commented:
@wrigri I've just checked the tests and they look good - thanks!

I still need to run them locally and make sure they pass with all the supported versions before +1'ing it though.

When the pull request is +1'ed, please attach final attach to the corresponding JIRA ticket.

Since this is a slightly larger patch, please also to file an ICLA ((https://libcloud.apache.org/contributing.html#contributing-bigger-changes)) if you haven't already.

wrigri committed:
Rename ex_list_volumes to list_volumes to align with LIBCLOUD-352

wrigri commented:
I sent a signed ICLA to Apache this morning.  In addition, I made a minor change based on the recent change to add list_volumes to the API.  Let me know if you find any additional issues that need to be addressed.

Kami committed:
Fix Python 2.5 compatibility issues.

Kami committed:
Fix a typo.

Kami commented:
I've opened a PR which fixes some 2.5 compatibility issues - https://github.com/wrigri/libcloud/pull/1

Once those issues are addresses, please attach a final patch to the JIRA ticket. I will merge it as soon as I receive a notification about ICLA filling from the ASF secretary.

wrigri committed:
Merge pull request #1 from Kami/LIBCLOUD-266_GCE_Support_python_2.5_compatibility_fixes

Fix Python 2.5 compatibility issues.

Kami commented:
This has been merged. You can close this pull request now.

Thanks!

Kami commented:
@wrigri Hey, I just noticed Google added support for load balancing*. Do you also plan to add support for it to libcloud?

It would be great if you could :)
- http://googlecloudplatform.blogspot.com/2013/08/google-cloud-platform-enhances-compute_7.html
- https://developers.google.com/compute/docs/load-balancing/

wrigri commented:
Yes, load balancing support is in progress and I should have a pull request out for that in the near future.  In addition, I have some misc enhancements and bug fixes in the compute engine driver coming shortly.

Kami commented:
@wrigri Awesome. :)
